dot_config/doom/config.org
* Resources
https://tecosaur.github.io/emacs-config/config.html
https://dotdoom.rgoswami.me/config
* Doom Keybinds
Set the leader to spacemacs
#+Begin_src emacs-lisp
(setq doom-localleader-key ",")
#+end_src
* Doom
There are two ways to load a theme. Both assume the theme is installed and
available. You can either set `doom-theme' or manually load a theme with the
`load-theme' function. This is the default:

#+begin_src emacs-lisp
(setq doom-theme 'doom-alabaster
      fancy-splash-image (concat doom-private-dir "logo/doom.png"))
#+end_src



This determines the style of line numbers in effect. If set to `nil', line
numbers are disabled. For relative line numbers, set this to `relative'.
#+begin_src emacs-lisp
(setq display-line-numbers-type `t)
#+end_src
* Evil-mode
#+begin_src emacs-lisp
(after! evil
  (setq evil-shift-round nil)
  (setq evil-respect-visual-line-mode t))
#+end_src
* Word Wrap
#+begin_src emacs-lisp
(+global-word-wrap-mode +1)
#+end_src
* Which Key
Set the whick key opening delay to  0.5 seconds
#+begin_src emacs-lisp
(setq which-key-idle-delay 0.5)
#+end_src

replace evil- in which-key entries to make em easier to read
#+begin_src emacs-lisp
(setq which-key-allow-multiple-replacements t)
(after! which-key
  (pushnew!
   which-key-replacement-alist
   '(("" . "\\`+?evil[-:]?\\(?:a-\\)?\\(.*\\)") . (nil . "◂ \\1"))
   '(("\\`g s" . "\\`evilem--?motion-\\(.*\\)") . (nil . "◃ \\1"))
   ))

#+end_src
* COMMENT Neoscroll
#+begin_src emacs-lisp
(use-package! neoscroll
  :config
  ;; Configure smooth scrolling parameters
  (setq neoscroll-easing 'cubic)
  (setq neoscroll-scroll-duration 0.15)
  (setq neoscroll-page-duration 0.25) 
  (setq neoscroll-line-duration 0.025)
  (setq neoscroll-line-step 3)  ; Scroll 3 lines with C-y/C-e
  
  ;; Enable neoscroll mode
  (neoscroll-mode 1))
#+end_src
* COMMENT Org Sliced Image
#+begin_src emacs-lisp
(after! org
  (org-sliced-images-mode 1))
#+end_src
* Org Mode
Set the directory
#+begin_src emacs-lisp

(add-hook 'org-mode-hook 'turn-on-org-cdlatex)
(add-hook 'LaTeX-mode-hook 'turn-on-cdlatex)
(add-hook 'org-mode-hook 'org-latex-preview-mode)

(defun enable-word-processor-minor-modes ()
  (setq line-spacing 0.15)
  ;;(pixel-scroll-precision-mode)
  (smartparens-mode)
  (visual-line-mode))
;;  (add-hook 'window-size-change-functions 'org-image-resize)
;;(add-hook 'text-mode-hook 'enable-word-processor-minor-modes)
;;(add-hook 'org-mode-hook 'org-appear-mode)
(add-hook 'org-mode-hook 'org-roam-db-autosync-enable)
(remove-hook 'org-mode-hook 'org-modern-mode)
(after! org
  (setq org-latex-preview-mode-display-live t)
  (setq org-directory "~/notes/org-mode/")
  (setq org-roam-directory "~/notes/org-mode/")
  (setq org-habit-show-habits-only-for-today nil)
  (setq org-agenda-files (directory-files-recursively "~/notes/org-mode/" "\\.org$"))
  (setq org-pretty-entities t)
  (setq org-hide-emphasis-markers t)
  (setq org-startup-folded 'content)
  ;; (setq org-startup-with-inline-images t)
  (setq org-startup-with-link-previews t)
  (setq org-startup-with-latex-preview t))
  ;; (setq org-ellipsis " ")) ;; folding symbol
#+end_src
** Org Writeroom
*** Setting the variable fonts
#+begin_src emacs-lisp
(setq doom-font (font-spec :family "IntoneMono Nerd Font Mono" :size 18)
      doom-variable-pitch-font (font-spec :family "Vollkorn"))
#+end_src
*** COMMENT Setting the fixed fonts
#+begin_src emacs-lisp
(setq doom-font (font-spec :family "Iosevka Nerd Font Mono" :size 18)
      doom-variable-pitch-font (font-spec :family "Iosevka Nerd Font Mono"))
#+end_src
***  Writeroom Mode
#+begin_src emacs-lisp
(after! writeroom-mode
  (setq +zen-text-scale 0)
  (setq writeroom-width 250)
  )
(add-hook 'org-mode-hook '+zen/toggle)
#+end_src
** Disable line numbers in org
#+begin_src emacs-lisp
(after! org
  (add-hook 'org-mode-hook #'doom-disable-line-numbers-h))
#+end_src

** bind j and k to evil-visual-line up and down.
#+begin_src emacs-lisp
(after! evil
  (map! :nv "j" #'evil-next-visual-line
        :nv "k" #'evil-previous-visual-line
        :nv "<down>" #'evil-next-visual-line
        :nv "<up>" #'evil-previous-visual-line))
#+end_src
** COMMENT Run buffers in org-mode by default:
#+begin_src emacs-lisp
(setq-default major-mode 'org-mode)
#+end_src
** Org Keybinds

I want the following keybinds to be added that i've noticed
The *insert* maps
The *text* maps
Comment lines
ctrl c ctrl c
copy and clipboard.
display org roam tags in node search
Other stuff- daemon mode,
smartparens equivalency (default +smartparens under doom config is NOT all you need, bind sp-sexp-up to tab)
** Org Appear
#+begin_src emacs-lisp
(after! org
  (setq org-appear-autolinks t)
  (setq org-appear-autoentities t)
  (setq org-appear-autosubmarkers t)
  (setq org-appear-autokeywords t)
  (setq org-appear-inside-latex t)
  )

;; It's cool to have appear only work in insert mode, gonna leave in automatic for now
;; (setq org-appear-trigger 'manual)
;; (add-hook 'org-mode-hook (lambda ()
;;                           (add-hook 'evil-insert-state-entry-hook
;;                                     #'org-appear-manual-start
;;                                     nil
;;                                     t)
;;                           (add-hook 'evil-insert-state-exit-hook
;;                                     #'org-appear-manual-stop
;;                                     nil
;;                                     t)))
#+end_src
** Org Download
#+begin_src emacs-lisp
(after! org-download
  (setq org-download-method 'directory)
  (setq org-download-image-dir (concat org-directory ".attach/" ))
  (setq org-download-link-format "[[file:%s]]\n"
        org-download-abbreviate-filename-function #'file-relative-name)
  (setq org-download-link-format-function #'org-download-link-format-function-default)
  (setq org-download-disable-id-create t)
)
#+end_src

#+RESULTS:
: t

Add an advice to supress org-get-id call on org-download-clipboard
#+begin_src emacs-lisp
(after! org
(advice-remove 'org-download-clipboard 'org-id-get-create))
#+end_src
** Org-roam
Fix that annoying sqlitep nil error:
#+begin_src emacs-lisp
(defun patch/emacsql-close (connection &rest args)
  "Prevent calling emacsql-close if connection handle is nil."
  (when (oref connection handle)
    t))

(advice-add 'emacsql-close :before-while #'patch/emacsql-close)
#+end_src

*** Set org roam db location to synced notes folder
#+begin_src emacs-lisp
(after! org
  (setq org-roam-db-location "~/notes/org-mode/org.db"))
#+end_src
*** Org roam completions to only link at point, it's annoying otherwise
#+begin_src emacs-lisp
(after! org
  (setq org-roam-completion-functions (list #'org-roam-complete-link-at-point)))

#+end_src
** Org-indent
#+begin_src emacs-lisp
(after! org
  (setq org-indent-indentation-per-level 4))
#+end_src
** Search content folded headings
#+begin_src emacs-lisp
(after! org
  (setq org-fold-core-style 'overlays)
  (evil-select-search-module 'evil-search-module 'evil-search))
#+end_src
** Org Latex Preview Auto Mode
** Org Mode Image config
#+begin_src emacs-lisp
(after! org
  (setq org-image-actual-width 300)
  (setq org-image-max-width 300))
#+end_src
* Org Babel
** Org Mermaid
#+begin_src emacs-lisp
(after! ob-mermaid
  (setq ob-mermaid-cli-path "mmdc"))
#+end_src
** Org Babel List
#+begin_src emacs-lisp
(after! org
  (org-babel-do-load-languages
    'org-babel-load-languages
    '((mermaid . t)
      (emacs-lisp . t))))
#+end_src
* Org Export
** Ox clip clipboard command
#+begin_src emacs-lisp
(after! org
  (setq ox-clip-linux-cmd "cat %f | wl-copy"))
#+end_src
* Latex
#+begin_src emacs-lisp
(after! tex
  (setq TeX-electric-sub-and-superscript t)
  (setq +latex-viewers '(zathura)))
#+end_src
* Transparency
#+begin_src  emacs-lisp
(set-frame-parameter nil 'alpha-background 100)
#+end_src
* LSP
Set up eglot clangd
#+begin_src emacs-lisp
(set-eglot-client! 'cc-mode '("clangd" "-j=3" "--clang-tidy"))
#+end_src

* Projectile
#+begin_src emacs-lisp
  (setq projectile-cache-file (concat doom-cache-dir "projectile.cache")
        projectile-enable-caching (not noninteractive)
        projectile-indexing-method (if IS-WINDOWS 'native 'alien)
        projectile-known-projects-file (concat doom-cache-dir "projectile.projects")
        projectile-require-project-root nil
        projectile-globally-ignored-files '(".DS_Store" "Icon
" "TAGS")
        projectile-globally-ignored-file-suffixes '(".elc" ".pyc" ".o")
        projectile-ignored-projects '("~/" "/tmp"))
#+end_src
* Web languages
HTML Auto Preview
#+begin_src emacs-lisp
(after! web-mode
  (add-hook 'web-mode-hook #'htmlz-mode))
#+end_src
* COMMENT Typst
** add major mode
#+begin_src emacs-lisp
(add-to-list 'auto-mode-alist '("\\.typ\\'" . typst-ts-mode))
#+end_src
** ts mode
#+begin_src emacs-lisp

(add-hook! 'typst-ts-mode-hook 'eglot-ensure)
(after! typst-ts-mode
  (setq typst-ts-watch-options '("--open"))
  (setq typst-ts-mode-enable-raw-blocks-highlight t)
  )
#+end_src
** LSP
#+begin_src emacs-lisp
(set-eglot-client! 'typst-ts-mode '("tinymist"))
#+end_src
** Typst preview
#+begin_src emacs-lisp
(after! typst-ts
  (setq typst-preview-autostart t)
  (setq typst-preview-open-browser-automatically t)
  (setq typst-preview-browser "default")  ; this is the default option; other options are `eaf-browser' or `xwidget'.
  (setq typst-preview-invert-colors "never") ; invert colors depending on system theme
  (setq typst-preview-executable "tinymist") ; path to tinymist binary (relative or absolute)
  (setq typst-preview-partial-rendering t))   ; enable partial rendering 
#+end_src


dot_config/doom/config.el
(setq doom-localleader-key ",")

(setq doom-theme 'doom-alabaster
      fancy-splash-image (concat doom-private-dir "logo/doom.png"))

(setq display-line-numbers-type `t)

(after! evil
  (setq evil-shift-round nil)
  (setq evil-respect-visual-line-mode t))

(+global-word-wrap-mode +1)

(setq which-key-idle-delay 0.5)

(setq which-key-allow-multiple-replacements t)
(after! which-key
  (pushnew!
   which-key-replacement-alist
   '(("" . "\\`+?evil[-:]?\\(?:a-\\)?\\(.*\\)") . (nil . "◂ \\1"))
   '(("\\`g s" . "\\`evilem--?motion-\\(.*\\)") . (nil . "◃ \\1"))
   ))

(add-hook 'org-mode-hook 'turn-on-org-cdlatex)
(add-hook 'LaTeX-mode-hook 'turn-on-cdlatex)
(add-hook 'org-mode-hook 'org-latex-preview-mode)

(defun enable-word-processor-minor-modes ()
  (setq line-spacing 0.15)
  ;;(pixel-scroll-precision-mode)
  (smartparens-mode)
  (visual-line-mode))
;;  (add-hook 'window-size-change-functions 'org-image-resize)
;;(add-hook 'text-mode-hook 'enable-word-processor-minor-modes)
;;(add-hook 'org-mode-hook 'org-appear-mode)
(add-hook 'org-mode-hook 'org-roam-db-autosync-enable)
(remove-hook 'org-mode-hook 'org-modern-mode)
(after! org
  (setq org-latex-preview-mode-display-live t)
  (setq org-directory "~/notes/org-mode/")
  (setq org-roam-directory "~/notes/org-mode/")
  (setq org-habit-show-habits-only-for-today nil)
  (setq org-agenda-files (directory-files-recursively "~/notes/org-mode/" "\\.org$"))
  (setq org-pretty-entities t)
  (setq org-hide-emphasis-markers t)
  (setq org-startup-folded 'content)
  ;; (setq org-startup-with-inline-images t)
  (setq org-startup-with-link-previews t)
  (setq org-startup-with-latex-preview t))
  ;; (setq org-ellipsis " ")) ;; folding symbol

(setq doom-font (font-spec :family "IntoneMono Nerd Font Mono" :size 18)
      doom-variable-pitch-font (font-spec :family "Vollkorn"))

(after! writeroom-mode
  (setq +zen-text-scale 0)
  (setq writeroom-width 250)
  )
(add-hook 'org-mode-hook '+zen/toggle)

(after! org
  (add-hook 'org-mode-hook #'doom-disable-line-numbers-h))

(after! evil
  (map! :nv "j" #'evil-next-visual-line
        :nv "k" #'evil-previous-visual-line
        :nv "<down>" #'evil-next-visual-line
        :nv "<up>" #'evil-previous-visual-line))

(after! org
  (setq org-appear-autolinks t)
  (setq org-appear-autoentities t)
  (setq org-appear-autosubmarkers t)
  (setq org-appear-autokeywords t)
  (setq org-appear-inside-latex t)
  )

;; It's cool to have appear only work in insert mode, gonna leave in automatic for now
;; (setq org-appear-trigger 'manual)
;; (add-hook 'org-mode-hook (lambda ()
;;                           (add-hook 'evil-insert-state-entry-hook
;;                                     #'org-appear-manual-start
;;                                     nil
;;                                     t)
;;                           (add-hook 'evil-insert-state-exit-hook
;;                                     #'org-appear-manual-stop
;;                                     nil
;;                                     t)))

(after! org-download
  (setq org-download-method 'directory)
  (setq org-download-image-dir (concat org-directory ".attach/" ))
  (setq org-download-link-format "[[file:%s]]\n"
        org-download-abbreviate-filename-function #'file-relative-name)
  (setq org-download-link-format-function #'org-download-link-format-function-default)
  (setq org-download-disable-id-create t)
)

(after! org
(advice-remove 'org-download-clipboard 'org-id-get-create))

(defun patch/emacsql-close (connection &rest args)
  "Prevent calling emacsql-close if connection handle is nil."
  (when (oref connection handle)
    t))

(advice-add 'emacsql-close :before-while #'patch/emacsql-close)

(after! org
  (setq org-roam-db-location "~/notes/org-mode/org.db"))

(after! org
  (setq org-roam-completion-functions (list #'org-roam-complete-link-at-point)))

(after! org
  (setq org-indent-indentation-per-level 4))

(after! org
  (setq org-fold-core-style 'overlays)
  (evil-select-search-module 'evil-search-module 'evil-search))

(after! org
  (setq org-image-actual-width 300)
  (setq org-image-max-width 300))

(after! ob-mermaid
  (setq ob-mermaid-cli-path "mmdc"))

(after! org
  (org-babel-do-load-languages
    'org-babel-load-languages
    '((mermaid . t)
      (emacs-lisp . t))))

(after! org
  (setq ox-clip-linux-cmd "cat %f | wl-copy"))

(after! tex
  (setq TeX-electric-sub-and-superscript t)
  (setq +latex-viewers '(zathura)))

(set-frame-parameter nil 'alpha-background 100)

(set-eglot-client! 'cc-mode '("clangd" "-j=3" "--clang-tidy"))

(setq projectile-cache-file (concat doom-cache-dir "projectile.cache")
        projectile-enable-caching (not noninteractive)
        projectile-indexing-method (if IS-WINDOWS 'native 'alien)
        projectile-known-projects-file (concat doom-cache-dir "projectile.projects")
        projectile-require-project-root nil
        projectile-globally-ignored-files '(".DS_Store" "Icon
" "TAGS")
        projectile-globally-ignored-file-suffixes '(".elc" ".pyc" ".o")
        projectile-ignored-projects '("~/" "/tmp"))

(after! web-mode
  (add-hook 'web-mode-hook #'htmlz-mode))
